### PR TITLE
feat(enhancement): Max number of one Fleet

### PR DESCRIPTION
**Pull request by RisingLeaf**
-----------------------
**Feature:** This PR implements a feature shortly discussed on github

## Feature Details
This PR adds a third attrib to the fleet tag in system.
`fleet "name" peroid number`
So that one fleet can only appear up to `number` times in one system.

### CODE REVIEW

**Comment by EjoThims** [comment]
Is this max number of the fleet in the system at one time, or max number of times (presumably per day?) that this fleet will enter the system?

**Comment by thvk-net** [comment]
The max number of fleets that can exist in the system

**Comment by warp-core** [comment]
> Is this max number of the fleet in the system at one time, or max number of times (presumably per day?) that this fleet will enter the system?

I haven't gone all the way through, but it looks like the former.

File: source/System.h [hunk 1]
```diff
@@ -197,6 +208,8 @@ class System {
 	std::vector<Asteroid> asteroids;
 	const Sprite *haze = nullptr;
 	std::vector<RandomEvent<Fleet>> fleets;
+	std::vector<Counter> fleetCounters;
```
Comment (warp-core) [inline source/System.h:211]: Any reason this can't be an `std::map<int, int>`?

File: source/Fleet.cpp [hunk 2]
```diff
@@ -430,7 +430,10 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
 		if(flagship)
 			ship->SetParent(flagship);
 		else
+		{
 			flagship = ship;
+			ship->SetFleetId(fleetId);
```
Comment (warp-core) [inline source/Fleet.cpp:435]: If only the flagship has the fleetId, that means, if the flagship is gone, this fleet no longer contributes towards the cap, even if every other ship in the fleet is still present?
Is this the desired behaviour?

File: source/System.h (hunk 1 above)
Comment (thvk-net) [reply source/System.h:211]: Ah, no

File: source/Fleet.cpp (hunk 2 above)
Comment (thvk-net) [reply source/Fleet.cpp:435]: Yes

Comment (thvk-net) [reply source/Fleet.cpp:435]: I could not find a better solution

File: source/System.h (hunk 1 above)
Comment (thvk-net) [reply source/System.h:211]: I will do this tomorrow

File: source/Fleet.cpp (hunk 2 above)
Comment (warp-core) [reply source/Fleet.cpp:435]: I suppose stepping through every ship in a system to check if there are any ships left with the fleetId of one being removed wouldn't be great for performance.

File: source/System.h (hunk 1 above)
Comment (thvk-net) [reply source/System.h:211]: No wait I am doing it now

File: source/Fleet.cpp (hunk 2 above)
Comment (Terin) [reply source/Fleet.cpp:435]: Most uses I can think of for this feature would be to limit single ship fleets, or limit a particular flagship with normal escorts. I think that's an acceptable limitation.

Comment (thvk-net) [reply source/Fleet.cpp:435]: > I suppose stepping through every ship in a system to check if there are any ships left with the fleetId of one being removed wouldn't be great for performance.

Not really yes, i mean its bad for battles but when you are not in combat all ships of one fleet should leave simultan

File: source/System.h (hunk 1 above)
Comment (thvk-net) [reply source/System.h]: done

File: source/Fleet.cpp (hunk 2 above)
Comment (thvk-net) [reply source/Fleet.cpp:435]: > Most uses I can think of for this feature would be to limit single ship fleets, or limit a particular flagship with normal escorts. I think that's an acceptable limitation.

Well if you do not care about accuracy then you can apply it even to larger fleets

File: source/System.h (hunk 1 above)
Comment (thvk-net) [reply source/System.h]: I just like structs

File: source/Fleet.cpp (hunk 2 above)
Comment (thvk-net) [reply source/Fleet.cpp:435]: New idea:
You could check for the escorts of the dead ship make one of the escorts parent over the others and write the fleetId to it

Comment (thvk-net) [reply source/Fleet.cpp:435]: While this is faster then chhecking all ships it is still a huge process

**Comment by warp-core** [comment]
@Amazinite @Zitchas
Should the fleet cap only consider the flagship of a fleet (if that ship leaves/is destroyed, its instance of the fleet no longer contributes towards the cap, even if every other ship is still present) or should very ship be checked?
This code currently only considers the flagship, checking every ship could be a lot of extra computation.
We could limit that by having each system set a flag as to whether or not it has any fleet caps, but systems with fleet caps would still be checking many of the ships present every time one leaves or is destroyed, this would likely be especially problematic when fighters are involved (see: Mesuket). I think better solutions would probably require larger refactors @thvk-net could probably do, but would take longer and require more reviewing.

For my part, I think just considering the flagship is an acceptable solution.

File: source/Fleet.cpp (hunk 2 above)
Comment (EjoThims) [reply source/Fleet.cpp:435]: I agree with Terin and Warp that checking for flagship should be sufficient.

Doubly so if the alternative is a big performance hit.

**Comment by thvk-net** [comment]
It would be nice if someone tested this because i am not sure if it is working how i want it to work

**Comment by thvk-net** [comment]
If someone can approve that this works i will make it a real pr

File: source/Fleet.cpp (hunk 2 above)
Comment (thvk-net) [reply source/Fleet.cpp:435]: Another idea: we make a vector with pointers to all ships of the fleet. All we have to do then is checking if they are nullptr

Comment (thvk-net) [reply source/Fleet.cpp:435]: But i dont know if that would work

**Comment by thvk-net** [comment]
I have to say that I wont be able to continue working on this for the next few weeks.

**Comment by Hurleveur** [comment]
Why are you using fleetIds instead of just having pointers to the fleets?

I agree with just considering the flagship for now, we can always expand that behaviour later on.
That said maybe we shouldn't check if its alive every frame and instead have the engine change what needs to be changed on its death? Whenever a ship is killed we check if it was part of one such fleet.
Doing that, we could easily expand the behaviour to concern all ships, I believe (in another PR, ofc).

**Comment by Zitchas** [comment]
I'm good with this simply checking the flagship. That's generally the most important part of the fleet anyway.

That being said, I do like the sound of Hurleveur's suggestion to flip it around so instead of having it being a continuous check to see if the ship is alive; it is instead something that takes place on its death.

Pseudo code example:
1. When the fleet SuperCruiser is spawned, each ship adds +1 to the SuperCruiserPresent variable.
2. Each time a ship from that fleet dies, it does -1 to the SuperCruiserPresent variable.
3. Each time the SuperCruiser fleet spawn check comes up, it just looks at whether or not SuperCruiserPresent is =< ((SuperCruiser fleet count -1) * number of ships in the fleet)

Just an idea for a future refinement of this, anyway. I'm perfectly happy to just have it go by flagship for now.

**Comment by warp-core** [comment]
If I read it right, the check was only happening when a ship is being removed, not continuously.

**Comment by Hurleveur** [comment]
Oh right, I didn't see because of github only showing such a minimum scope, sorry (and me going too quickly too), sorry.

**Comment by 1010todd** [comment]
> It would be nice if someone tested this because i am not sure if it is working how i want it to work

Apparently not, with my test plugin below upon departing I see 3-4 fleets in the system and after they jumped out/landed nothing else appears. 

(Plugin below replace Rutilicus fleets with one custom fleet with one Bactrian each at 100 frequency.)

[zfleet limit test.zip](https://github.com/endless-sky/endless-sky/files/9438019/zfleet.limit.test.zip)

**Comment by Hurleveur** [comment]
Oh right this code only takes into account the ship dying, not leaving the system, or landing, and thus "normally" disappearing?

I think they just don't get replaced by the engine when this happens, and the old vector is cleared. Would be worth checking if landing refreshes the count or not.

**Comment by warp-core** [comment]
> Why are you using fleetIds instead of just having pointers to the fleets?

I don't think Engine has access to the instantiated fleet object? I think doing it that way would require some larger refactoring which I don't think is necessary at this stage.

It does look like `ShouldBeRemoved()` will only be true if the ship has been destroyed, not if it left the system, which is inconvenient.
That does raise the question of how a ship leaving the system should be handled.
If it leaving reduces the number of instances of a fleet below the maximum, and another of that fleet is spawned, would we want to prevent that ship from returning to the system in addition to preventing the spawning of additional instances of that fleet?

**Comment by warp-core** [comment]
I don't think this is actually updating the counter in System properly, either.
I'm going to open a PR to the branch for this that should fix it.

**Comment by warp-core** [comment]
An additional issue that occurs to me: fleetIds seem to only be unique within a system, so if a ship moves to a different system, it may prevent fleets spawning in that system that should be allowed to.

**Comment by thvk-net** [comment]
> An additional issue that occurs to me: fleetIds seem to only be unique within a system, so if a ship moves to a different system, it may prevent fleets spawning in that system that should be allowed to.

Ah no, because fleets are unique in each system with unique ids assigned to them.

**Comment by thvk-net** [comment]
> That does raise the question of how a ship leaving the system should be handled.
> If it leaving reduces the number of instances of a fleet below the maximum, and another of that fleet is spawned, would we want to prevent that ship from returning to the system in addition to preventing the spawning of additional instances of that fleet?

Good question, I need an answer

**Comment by thvk-net** [comment]
> Pseudo code example:
> 
> 1. When the fleet SuperCruiser is spawned, each ship adds +1 to the SuperCruiserPresent variable.
> 
> 2. Each time a ship from that fleet dies, it does -1 to the SuperCruiserPresent variable.
> 
> 3. Each time the SuperCruiser fleet spawn check comes up, it just looks at whether or not SuperCruiserPresent is =< ((SuperCruiser fleet count -1) * number of ships in the fleet)

Wont work. For this to work all fleet variants would need to have the same number of ships

**Comment by thvk-net** [comment]
Because this came dead now, I am asking again: how should ships out of the system be handled, as they might return.

**Comment by thvk-net** [comment]
Sorry i did not want to make this a real pr

**Comment by Hurleveur** [comment]
> Because this came dead now, I am asking again: how should ships out of the system be handled, as they might return.

Lets do this the simple way: only consider the amount of flagships in the system, if we have more than that don't spawn anything, else we may spawn some.
If they come back, its not big deal, we just won't make new fleets spawn whilst they're there.

**Comment by thvk-net** [comment]
> > Because this came dead now, I am asking again: how should ships out of the system be handled, as they might return.
> 
> 
> 
> Lets do this the simple way: only consider the amount of flagships in the system, if we have more than that don't spawn anything, else we may spawn some.
> 
> If they come back, it's no big deal, just won't make new fleets spawn whilst they're there. I don't think this needs to be perfect.

Ok

**Comment by RisingLeaf** [comment]
I have to say that I wont be able to work on this in the coming months. So if someone wants this feature, they have to find someone else

**Comment by Zitchas** [comment]
I'm sorry that you weren't able to continue this. This would be particularly useful in combination with the fleet conditions being worked on in #7947 .

Is there any chance that anyone else may be able to pick this up?

**Pull request closed without merging**